keyring/backends/Windows.py

```diff
@@ -88,79 +301,152 @@ def priority(cls):
             # It wasn't found so attempt to get it with the compound name
-            res = self._get_password(self._compound_name(username, service))
-        if not res:
+            creds = self._get_password(self._compound_name(username, service))
+        if not creds:
             return None
-        return res.value
+        password = creds.value
+        return password
 
-    def _get_password(self, target):
+    def _get_password_inner(self, target):
         try:
-            res = win32cred.CredRead(
-                Type=win32cred.CRED_TYPE_GENERIC, TargetName=target
-            )
-        except pywintypes.error as e:
-            e = OldPywinError.wrap(e)
-            if e.winerror == 1168 and e.funcname == 'CredRead':  # not found
+            res = CredRead(Type=CRED_TYPE_GENERIC, TargetName=target)
+        except CredError as e:
+            if e.winerror == 1168 and e.funcname == "CredRead":  # not found
                 return None
             raise
         return DecodingCredential(res)
 
+    def _get_password(self, target):
+        creds = self._get_password_inner(target)
+        if creds and ATTRIBUTE_KEYWORD in creds and creds[ATTRIBUTE_KEYWORD]["max_shards"] > 1:
+            for i in range(1, creds[ATTRIBUTE_KEYWORD]["max_shards"]):
+                shard = self._get_password_inner(TARGET_SHARD.format(target=target, n=i))
+                if not shard:
+                    log.critical(f"expected shard {i} not found; {creds[ATTRIBUTE_KEYWORD]}")
+                    return creds
+                else:
+                    creds["CredentialBlob"] += shard["CredentialBlob"]
+        return creds
+
     def set_password(self, service, username, password):
         existing_pw = self._get_password(service)
         if existing_pw:
-            # resave the existing password using a compound target
-            existing_username = existing_pw['UserName']
-            target = self._compound_name(existing_username, service)
-            self._set_password(
-                target,
-                existing_username,
-                existing_pw.value,
-            )
+            existing_username = existing_pw["UserName"]
+            if existing_username != username:
+                # resave the existing password using a compound target
+                target = self._compound_name(existing_username, service)
+                self._set_password(target, existing_username, existing_pw.value)
         self._set_password(service, username, str(password))
 
     def _set_password(self, target, username, password):
-        credential = dict(
-            Type=win32cred.CRED_TYPE_GENERIC,
-            TargetName=target,
-            UserName=username,
-            CredentialBlob=password,
-            Comment="Stored using python-keyring",
-            Persist=self.persist,
-        )
-        win32cred.CredWrite(credential, 0)
+        pwd_utf8_bytes = password.encode("utf-8")
```

Comment:
Probably the biggest concern is this new default encoding. The reason keyring defaulted to UTF-16 was because that's the encoding that other tools (except Electron) appear to use UTF-16 by default. See #438 for more details. I'd like for the stored passwords to be interoperable where possible (i.e. a password stored by keyring should be readable by other programs). Obviously, they won't be interoperable if they are sharded, but this change alters the default encoding as well. And unfortunately, there's no standard here, just a popular convention. I don't think it will be advisable to change the default encoding to utf-8 until there's a substantial move to support that format in other tools. This change would also be backward incompatible in that it would make new passwords incompatible with older keyrings prior to support for UTF-8 decoding.

Reply:
I'm fine sticking with the UTF-16 encoding.  I only switched because I noticed that, for token storage, we were wasting half the bytes in the record.  A typical JWT fits in a single record under UTF-8, but spills to two sharded records in UTF-16.  In that case (a typical JWT), either way other programs on Windows aren't going to be able to read it, unless they support the proposed attribute scheme. 

But, yes, a password that would fit under UTF-16 would become non-backwards compatible with my code, so I will switch back to UTF-16.